rfd/0019-event-iteration-api.md

```diff
@@ -0,0 +1,114 @@
+
+```proto
+rpc GetEvents(GetEventsRequest) returns (Events);
+```
+
+```proto
+rpc GetSessionEvents(GetSessionEventsRequest) returns (Events);
+```
+
+### Implementation details 3.3
+
+#### Auth service 3.3.1
+
+Handling of these new GRPC endpoints will be implemented in [lib/auth/grpcserver.go](https://github.com/gravitational/teleport/blob/master/lib/auth/grpcserver.go). `auth.Client.SearchEvents` and `auth.Client.SearchSessionEvents` will be refactored with pagination support implemented with the same startkey/lastkey interface as seen in the GRPC API. These methods are what the GRPC handlers call. The current HTTP API implementation will be slightly altered to deal with these changes but will remain identicatal in functionality, simlpy not exposing the new internal pagination support.
+
+#### Backends 3.3.2
+
+Performed some investigative work on how we can implement this pagination support in the different backends. The standard solution that works with the current backend interface is to simply query all events in a given date range from the backend and then filter out the paginated section using the given start key and limit. This is however something we want to avoid as it involves a lot of uneeded processing on every request by the authentication server.
+
+Pagination support will needed to be added to the `IAuditLog` interface defined in `lib/events/api.go`. This involves adding a start key function parameter and an optional last key parameter in the return value of the `SearchEvents` and `SearchSessionEvents` methods.
+
+Pagination in DynamoDB seems possible to implement using start keys (partition + sort key) and global indexes for requesting only the paginated events and should be compatible with our current format.
+
+Not an expert on the format used for filelog but since it is a plain sequence of JSON documents, implementing efficient pagination seems difficult. On this backend we might need to fall back to loading all events into memory and then sorting out the events that fall within the range of events requested.
+
+Firestore seems to have good support for pagination with it's query cursors and document snapshots which allow us to define a range between startkey and startkey + limit. This means we can efficiently query a subset of events.
+
+Issue [#5885](https://github.com/gravitational/teleport/pull/5885) about implementing a MySQL event backend is also relevant to this discussion. If we were to directly implemented a MySQL compatible backend, we could perform pagination by giving each event a 64 bit integer number that increases sequentially. Event search would then be performed by filtering first by date and then selecting a range of event ID's. With a B-Tree index on dates and event ID's this would allow efficient pagination. If we go with an API/plugin based approach as @knisbet suggested, we could effectively allow for the same thing by exposing a search endpoint that in turn does this date and event ID filtering.
```

Comment:
I would push those out of scope of this PR, because it's not clear if we actually are going to support MySQL as a part of this RFD implementation (I assume not)

Reply:
Yeah for sure, just wanted to mention it and how we could do it but I could also remove it if we should address this separately later.

Reply:
Removed